controllers/clustercache/cluster_accessor.go

```diff
@@ -0,0 +1,492 @@
+// consecutiveFailures) of the clusterAccessor.
+type clusterAccessorLockedHealthCheckingState struct {
+	// lastProbeTimestamp is the time when the health probe was executed last.
+	lastProbeTimestamp time.Time
+
+	// lastProbeSuccessTimestamp is the time when the health probe was successfully executed last.
+	lastProbeSuccessTimestamp time.Time
+
+	// consecutiveFailures is the number of consecutive health probe failures.
+	consecutiveFailures int
+}
+
+// newClusterAccessor creates a new clusterAccessor.
+func newClusterAccessor(cluster client.ObjectKey, clusterAccessorConfig *clusterAccessorConfig) *clusterAccessor {
+	return &clusterAccessor{
+		cluster: cluster,
+		config:  clusterAccessorConfig,
+	}
+}
+
+// Connected returns true if there is a connection to the workload cluster, i.e. the clusterAccessor has a
+// client, cache, etc.
+func (ca *clusterAccessor) Connected(ctx context.Context) bool {
+	ca.rLock(ctx)
+	defer ca.rUnlock(ctx)
+
+	return ca.lockedState.connection != nil
+}
+
+// Connect creates a connection to the workload cluster, i.e. it creates a client, cache, etc.
+//
+// This method will only be called by the ClusterCache reconciler and not by other regular reconcilers.
+// Controller-runtime guarantees that the ClusterCache reconciler will never reconcile the same Cluster concurrently.
+// Thus, we can rely on this method not being called concurrently for the same Cluster / clusterAccessor.
+// This means we don't have to hold the lock when we create the client, cache, etc.
+// It is extremely important that we don't lock during connection creation, so that other methods like GetClient (that
+// use a read lock) can return immediately even when we currently create a client. This is necessary to avoid blocking
+// other reconcilers (e.g. the Cluster or Machine reconciler).
+func (ca *clusterAccessor) Connect(ctx context.Context) (retErr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if ca.Connected(ctx) {
+		log.V(6).Info("Skipping connect, already connected")
+		return nil
+	}
+
+	log.Info("Connecting")
+
+	// Creating clients, cache etc. is intentionally done without a lock to avoid blocking other reconcilers.
+	connection, err := ca.createConnection(ctx)
+
+	ca.lock(ctx)
+	defer ca.unlock(ctx)
+
+	defer func() {
+		if retErr != nil {
+			log.Error(retErr, "Connect failed")
+			ca.lockedState.lastConnectionCreationErrorTimestamp = time.Now()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	log.Info("Connected")
+
+	// Only generate the clientCertificatePrivateKey once as there is no need to regenerate it after disconnect/connect.
+	// Note: This has to be done before setting connection, because otherwise this code wouldn't be re-entrant if the
+	// private key generation fails because we check Connected above.
+	if ca.lockedState.clientCertificatePrivateKey == nil {
+		log.V(6).Info("Generating client certificate private key")
+		clientCertificatePrivateKey, err := certs.NewPrivateKey()
```

Comment:
Question: As this is propably the most time-consuming thing we do here:

Should we do this in a way that we are not locking the state?
But totally fine to me to keep as is if this seems to be okay also for scale already now (which I think it is, because we only do this the first time, and during that time there is no connection there yet which can be used by consumers of the ClusterCache)

Reply:
Yup this is basically run exactly once after we created the first connection.

While it takes the most time, we're talking about < 1 second.

I was thinking about the same, but it would also have required an additional read lock to check if the clientCertificatePrivateKey was already generated before.

So I thought the additional delay once is maybe better than the permanent additional read lock (and definitely slightly simpler)

Reply:
I think because it was the next best thing from generating a key at every connection to etcd (you know, incremental improvements).

If we want to improve Key handling further, I would prefer doing it in a follow up PR
(e.g. might be we can have only one key, and rotate it ~frequently

Reply:
I'm not sure why the private key would have to be rotated to be honest, but let's have that discussion separately (e.g. we also never rotate the private key of the 100 year Kubernetes CA we use for wl clusters)